pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java

```diff
@@ -0,0 +1,204 @@
+        }
+    }
+
+    @Override
+    public void clear() {
+        numMessagesInBatch = 0;
+        currentBatchSizeBytes = 0;
+        batches = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return batches.isEmpty();
+    }
+
+    @Override
+    public void handleException(Exception ex) {
+        try {
+            // Need to protect ourselves from any exception being thrown in the future handler from the application
+            batches.forEach((k, v) -> v.firstCallback.sendComplete(ex));
+        } catch (Throwable t) {
+            log.warn("[{}] [{}] Got exception while completing the callback", topicName, producerName, t);
+        }
+        batches.forEach((k, v) -> ReferenceCountUtil.safeRelease(v.batchedMessageMetadataAndPayload));
+        clear();
+    }
+
+    private ProducerImpl.OpSendMsg createOpSendMsg(KeyBasedPart keyBasedPart) throws IOException {
+        ByteBuf encryptedPayload = producer.encryptMessage(keyBasedPart.messageMetadata, keyBasedPart.getCompressedBatchMetadataAndPayload());
+        final int numMessagesInBatch = keyBasedPart.messages.size();
+        long currentBatchSizeBytes = 0;
+        for (MessageImpl<?> message : keyBasedPart.messages) {
+            currentBatchSizeBytes += message.getDataBuffer().readableBytes();
+        }
+        keyBasedPart.messageMetadata.setNumMessagesInBatch(numMessagesInBatch);
+        ByteBufPair cmd = producer.sendMessage(producer.producerId, keyBasedPart.sequenceId, numMessagesInBatch,
+                keyBasedPart.messageMetadata.build(), encryptedPayload);
+
+        ProducerImpl.OpSendMsg op = ProducerImpl.OpSendMsg.create(keyBasedPart.messages, cmd, keyBasedPart.sequenceId, keyBasedPart.firstCallback);
+
+        if (encryptedPayload.readableBytes() > ClientCnx.getMaxMessageSize()) {
+            cmd.release();
+            if (op != null) {
+                op.callback.sendComplete(new PulsarClientException.InvalidMessageException(
+                        "Message size is bigger than " + ClientCnx.getMaxMessageSize() + " bytes"));
+                op.recycle();
+            }
+            return null;
+        }
+        op.setNumMessagesInBatch(numMessagesInBatch);
+        op.setBatchSizeByte(currentBatchSizeBytes);
+        return op;
+    }
+
+    @Override
+    public List<ProducerImpl.OpSendMsg> createOpSendMsgs() throws IOException {
+        List<ProducerImpl.OpSendMsg> result = new ArrayList<>();
+        List<KeyBasedPart> list = new ArrayList<>(batches.values());
+        list.sort(((o1, o2) -> ComparisonChain.start()
+                .compare(o1.sequenceId, o2.sequenceId)
+                .result()));
+        for (KeyBasedPart keyBasedPart : list) {
+            ProducerImpl.OpSendMsg op = createOpSendMsg(keyBasedPart);
+            if (op != null) {
+                result.add(op);
+            }
+        }
+        return result;
+    }
+
+    private String peekKey(MessageImpl<?> msg) {
```

Comment:
`peek` sounds like an action to retrieve a key from a queue. why not just call it `getKey`?